lgc/elfLinker/NullFragmentShader.cpp

```diff
@@ -0,0 +1,74 @@
+ *
+ *  Copyright (c) 2021 Google LLC All Rights Reserved.
+ *
+ *  Permission is hereby granted, free of charge, to any person obtaining a copy
+ *  of this software and associated documentation files (the "Software"), to deal
+ *  in the Software without restriction, including without limitation the rights
+ *  to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ *  copies of the Software, and to permit persons to whom the Software is
+ *  furnished to do so, subject to the following conditions:
+ *
+ *  The above copyright notice and this permission notice shall be included in all
+ *  copies or substantial portions of the Software.
+ *
+ *  THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ *  IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ *  FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ *  AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ *  LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ *  OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ *  SOFTWARE.
+ *
+ **********************************************************************************************************************/
+/**
+ ***********************************************************************************************************************
+ * @file  NullFragmentShader.cpp
+ * @brief LGC source file: The class to generate the null fragment shader when linking.
+ ***********************************************************************************************************************
+ */
+
+#include "NullFragmentShader.h"
+#include "lgc/patch/FragColorExport.h"
+#include "lgc/patch/Patch.h"
+#include "lgc/state/TargetInfo.h"
+#include "llvm/Target/TargetMachine.h"
+
+using namespace lgc;
+using namespace llvm;
+
+// =====================================================================================================================
+// Generate the IR module for the null fragment shader
+//
+// @returns : The module containing the null fragment shader.
+Module *NullFragmentShader::generate() {
+  Module *module = generateEmptyModule();
+  Function *entryPoint = FragColorExport::generateNullFragmentShader(*module, getGlueShaderName());
+  addDummyExportIfNecessary(entryPoint);
+  return module;
+}
+
+// =====================================================================================================================
+// Adds a dummy export to the entry point if it is needed.
+//
+// @param [in/out] entryPoint : The function in which to add the dummy export.
+void NullFragmentShader::addDummyExportIfNecessary(Function *entryPoint) const {
+  if (m_lgcContext->getTargetInfo().getGfxIpVersion().major < 10) {
+    auto ret = cast<ReturnInst>(entryPoint->back().getTerminator());
+    BuilderBase builder(ret);
+    FragColorExport::addDummyExport(builder);
```

Comment:
I know this is what the old code did, but I think it is unnecessary. I tried a change internally where I made the null FS empty, relying on the back-end to insert a dummy export if the HW requires it. It passed our internal testing.

Reply:
I think I will leave it for now.  I don't know what you mean by "old code".  The dummy export in a regular FS is added here:

https://github.com/GPUOpen-Drivers/llpc/blob/3efbccc2a85df854db0c10cb4141efabc26ac799/lgc/patch/FragColorExport.cpp#L858-L860

This is called from the `LowerFragColorExport`.  We do not have to generate the dummy export in `PatchNullFragShader` because `LowerFragColorExport` will be run on the shader afterwards.

However, `LowerFragColorExport` is not run by the glue shader pass manager.  This is why the `FragColorExport` functions are called manually when generating the color export shader.  I think we need to do the same here.

Reply:
Fair enough, best to do the same then. But I believe we could remove the dummy export completely in a future change; at least that worked in my testing a couple of weeks ago. It works because the AMDGPU back-end adds a dummy export anyway, but only if it thinks the HW needs it.